File: src/core/GraphBuilder.js
export class GraphBuilder {
  constructor() {
    this.graph = new Map();
  }

  addNode(id, properties) {
    if (!this.graph.has(id)) {
      this.graph.set(id, {
        properties,
        edges: new Map()
      });
    }
    return this;
  }

  addEdge(fromId, toId, relation) {
    if (this.graph.has(fromId) && this.graph.has(toId)) {
      this.graph.get(fromId).edges.set(toId, relation);
    }
    return this;
  }

  validate() {
    // 验证图结构的完整性和正确性
    for (const [nodeId, node] of this.graph) {
      if (!this.validateNode(nodeId, node)) {
        return false;
      }
    }
    return true;
  }

  validateNode(nodeId, node) {
    // 验证节点的属性和关系是否有效
    return node.properties && 
           typeof node.properties === 'object' &&
           node.edges instanceof Map;
  }

  build() {
    if (!this.validate()) {
      throw new Error('Invalid graph structure');
    }
    return this.graph;
  }
}

File: src/core/graph_builder.py
import networkx as nx

class GraphBuilder:
    def __init__(self):
        self.graph = nx.MultiDiGraph()
        
    def add_node(self, node_id: str, properties: dict) -> 'GraphBuilder':
        if node_id not in self.graph:
            self.graph.add_node(node_id, **properties)
        return self
        
    def add_edge(self, from_id: str, to_id: str, relation: str) -> 'GraphBuilder':
        if from_id in self.graph and to_id in self.graph:
            self.graph.add_edge(from_id, to_id, relation=relation)
        return self
        
    def validate(self) -> bool:
        return all(self._validate_node(node) for node in self.graph.nodes)
        
    def _validate_node(self, node) -> bool:
        return isinstance(self.graph.nodes[node], dict)
        
    def build(self) -> nx.MultiDiGraph:
        if not self.validate():
            raise ValueError("Invalid graph structure")
        return self.graph